### ISSUE TITLE
mbedtls_ssl_handshake failed

### ISSUE THREAD

**Issue by shokuie** [aws/aws-iot-device-sdk-embedded-C#529]
folks I've read most of the posts but could not sort out my problem so thought I can borrow your mind to solve it.

Using the mbed ported version of sdk (https://github.com/Avnet/aws-iot-device-sdk-mbed-c) and using it on nucleo-l476rg board. Compiling it using AC6 System workbench (SW4STM32). After step by step improvement i got stock here, you can find complete debug output in the following lines,  but the summary is this error:

`ERROR: iot_tls_connect L#280  failed
  ! mbedtls_ssl_handshake returned -0x10 (-16)`

I've seen some posts mentioning 0x10 return value is something about memory shortage but I don't believe it while the sample from project works fine and I only replaced my aws account and certificates, so most probably it's not a memory issue. 

Debug output : 

AWS ../examples/shadow_sample/shadow_sample.cpp Example.

AWS IoT SDK Version 3.0.0-

Shadow Init
FUNC_ENTRY: aws_iot_shadow_init L#76 (../src/aws_iot_shadow.c)
FUNC_ENTRY: aws_iot_mqtt_init L#209 (../src/aws_iot_mqtt_client.c)
FUNC_ENTRY: aws_iot_mqtt_set_connect_params L#139 (../src/aws_iot_mqtt_client.c)
FUNC_EXIT:  aws_iot_mqtt_set_connect_params L#169 Return Code : 0 (../src/aws_iot_mqtt_client.c)
FUNC_ENTRY: iot_tls_init L#109 (../platform/mbedtls/network_mbedtls_wrapper.cpp)
FUNC_ENTRY: _iot_tls_set_connect_params L#95 (../platform/mbedtls/network_mbedtls_wrapper.cpp)
FUNC_EXIT:  iot_tls_init L#122 Return Code : 0 (../platform/mbedtls/network_mbedtls_wrapper.cpp)
FUNC_EXIT:  aws_iot_mqtt_init L#281 Return Code : 0 (../src/aws_iot_mqtt_client.c)
FUNC_EXIT:  aws_iot_shadow_init L#103 Return Code : 0 (../src/aws_iot_shadow.c)
Shadow Connect
FUNC_ENTRY: aws_iot_shadow_connect L#111 (../src/aws_iot_shadow.c)
FUNC_ENTRY: aws_iot_mqtt_connect L#461 (../src/aws_iot_mqtt_client_connect.c)
FUNC_ENTRY: aws_iot_mqtt_get_client_state L#63 (../src/aws_iot_mqtt_client.c)
FUNC_EXIT:  aws_iot_mqtt_get_client_state L#68 Return Code : 1 (../src/aws_iot_mqtt_client.c)
FUNC_ENTRY: aws_iot_mqtt_set_client_state L#110 (../src/aws_iot_mqtt_client.c)
FUNC_ENTRY: aws_iot_mqtt_get_client_state L#63 (../src/aws_iot_mqtt_client.c)
FUNC_EXIT:  aws_iot_mqtt_get_client_state L#68 Return Code : 1 (../src/aws_iot_mqtt_client.c)
FUNC_EXIT:  aws_iot_mqtt_set_client_state L#135 Return Code : 0 (../src/aws_iot_mqtt_client.c)
FUNC_ENTRY: _aws_iot_mqtt_internal_connect L#391 (../src/aws_iot_mqtt_client_connect.c)
FUNC_ENTRY: aws_iot_mqtt_set_connect_params L#139 (../src/aws_iot_mqtt_client.c)
FUNC_EXIT:  aws_iot_mqtt_set_connect_params L#169 Return Code : 0 (../src/aws_iot_mqtt_client.c)
FUNC_ENTRY: iot_tls_connect L#140 (../platform/mbedtls/network_mbedtls_wrapper.cpp)
DEBUG:   iot_tls_connect L#163
Seed the random number generator...
DEBUG:   iot_tls_connect L#171 Load the IoT certificate ...
DEBUG:   iot_tls_connect L#183 Load the private key ...
DEBUG:   iot_tls_connect L#196 Load the CA root certificate ...
DEBUG:   iot_tls_connect L#208 done.
Setting up the SSL/TLS structure...
DEBUG:   iot_tls_connect L#229 Set the SSL BIO ...
FUNC_ENTRY: mbedtls_aws_init L#61 (../platform/mbedtls/network.cpp)
[EasyConnect] Using BG96
[EasyConnect] Connected to Network successfully
[EasyConnect] MAC address XX:00:20:00:20
[EasyConnect] IP address xxx.xxx.97.119
DEBUG:   iot_tls_connect L#260 Connecting to XXXXXXXXXXXXX-ats.iot.us-east-1.amazonaws.com/443...
FUNC_ENTRY: mbedtls_aws_connect L#83 (../platform/mbedtls/network.cpp)
FUNC_EXIT:  mbedtls_aws_connect L#94 Return Code : 0 (../platform/mbedtls/network.cpp)
DEBUG:   iot_tls_connect L#276

Perform the SSL/TLS handshake...

FUNC_ENTRY: mbedtls_aws_send L#192 (../platform/mbedtls/network.cpp)
FUNC_EXIT:  mbedtls_aws_send L#204 Return Code : 242 (../platform/mbedtls/network.cpp)
FUNC_ENTRY: mbedtls_aws_recv_timeout L#169 (../platform/mbedtls/network.cpp)
FUNC_ENTRY: mbedtls_aws_recv L#151 (../platform/mbedtls/network.cpp)
FUNC_EXIT:  mbedtls_aws_recv L#159 Return Code : 5 (../platform/mbedtls/network.cpp)
FUNC_EXIT:  mbedtls_aws_recv_timeout L#180 Return Code : 5 (../platform/mbedtls/network.cpp)
FUNC_ENTRY: mbedtls_aws_recv_timeout L#169 (../platform/mbedtls/network.cpp)
FUNC_ENTRY: mbedtls_aws_recv L#151 (../platform/mbedtls/network.cpp)
FUNC_EXIT:  mbedtls_aws_recv L#159 Return Code : 1415 (../platform/mbedtls/network.cpp)
FUNC_EXIT:  mbedtls_aws_recv_timeout L#180 Return Code : 1415 (../platform/mbedtls/network.cpp)
FUNC_ENTRY: mbedtls_aws_recv_timeout L#169 (../platform/mbedtls/network.cpp)
FUNC_ENTRY: mbedtls_aws_recv L#151 (../platform/mbedtls/network.cpp)
FUNC_EXIT:  mbedtls_aws_recv L#159 Return Code : 1396 (../platform/mbedtls/network.cpp)
FUNC_EXIT:  mbedtls_aws_recv_timeout L#180 Return Code : 1396 (../platform/mbedtls/network.cpp)
FUNC_ENTRY: mbedtls_aws_recv_timeout L#169 (../platform/mbedtls/network.cpp)
FUNC_ENTRY: mbedtls_aws_recv L#151 (../platform/mbedtls/network.cpp)
FUNC_EXIT:  mbedtls_aws_recv L#159 Return Code : 1420 (../platform/mbedtls/network.cpp)
FUNC_EXIT:  mbedtls_aws_recv_timeout L#180 Return Code : 1420 (../platform/mbedtls/network.cpp)
FUNC_ENTRY: mbedtls_aws_recv_timeout L#169 (../platform/mbedtls/network.cpp)
FUNC_ENTRY: mbedtls_aws_recv L#151 (../platform/mbedtls/network.cpp)
FUNC_EXIT:  mbedtls_aws_recv L#159 Return Code : 1167 (../platform/mbedtls/network.cpp)
FUNC_EXIT:  mbedtls_aws_recv_timeout L#180 Return Code : 1167 (../platform/mbedtls/network.cpp)
FUNC_ENTRY: _iot_tls_verify_cert L#75 (../platform/mbedtls/network_mbedtls_wrapper.cpp)
DEBUG:   _iot_tls_verify_cert L#77 Verify requested for (Depth 2):
DEBUG:   _iot_tls_verify_cert L#79 cert. version     : 3
serial number     : XX:XX:XX:XX:XX:XX:8C:0A:39:E2:F0:78:8A:43:E6:96:36:5B:CA
issuer name       : C=US, O=Amazon, CN=Amazon Root CA 1
subject name      : C=US, O=Amazon, CN=Amazon Root CA 1
issued  on        : 2015-05-26 00:00:00
expires on        : 2038-01-17 00:00:00
signed using      : RSA with SHA-256
RSA key size      : 2048 bits
basic constraints : CA=true
key usage         : Digital Signature, Key Cert Sign, CRL Sign

DEBUG:   _iot_tls_verify_cert L#82 This certificate has no flags
FUNC_EXIT:  _iot_tls_verify_cert L#88 Return Code : 0 (../platform/mbedtls/network_mbedtls_wrapper.cpp)
FUNC_ENTRY: _iot_tls_verify_cert L#75 (../platform/mbedtls/network_mbedtls_wrapper.cpp)
DEBUG:   _iot_tls_verify_cert L#77 Verify requested for (Depth 1):
DEBUG:   _iot_tls_verify_cert L#79 cert. version     : 3
serial number     : XX:XX:XX:XX:XX:XX:AC:77:DE:B2:53:32:5B:BC:99:8B:56:0D
issuer name       : C=US, O=Amazon, CN=Amazon Root CA 1
subject name      : C=US, O=Amazon, OU=Server CA 1B, CN=Amazon
issued  on        : 2015-10-22 00:00:00
expires on        : 2025-10-19 00:00:00
signed using      : RSA with SHA-256
RSA key size      : 2048 bits
basic constraints : CA=true, max_pathlen=0
key usage         : Digital Signature, Key Cert Sign, CRL Sign

DEBUG:   _iot_tls_verify_cert L#82 This certificate has no flags
FUNC_EXIT:  _iot_tls_verify_cert L#88 Return Code : 0 (../platform/mbedtls/network_mbedtls_wrapper.cpp)
FUNC_ENTRY: _iot_tls_verify_cert L#75 (../platform/mbedtls/network_mbedtls_wrapper.cpp)
DEBUG:   _iot_tls_verify_cert L#77 Verify requested for (Depth 0):
DEBUG:   _iot_tls_verify_cert L#79 cert. version     : 3
serial number     : XX:XX:XX:XX:XX:XX::BD:03:1B:F0:ED:3B:7A:E7:66
issuer name       : C=US, O=Amazon, OU=Server CA 1B, CN=Amazon
subject name      : CN=*.iot.us-east-1.amazonaws.com
issued  on        : 2019-07-02 00:00:00
expires on        : 2020-06-02 12:00:00
signed using      : RSA with SHA-256
RSA key size      : 2048 bits
basic constraints : CA=false
subject alt name  : iot.us-east-1.amazonaws.com, *.iot.us-east-1.amazonaws.com
key usage         : Digital Signature, Key Encipherment
ext key usage     : TLS Web Server Authentication, TLS Web Client Authentication

DEBUG:   _iot_tls_verify_cert L#82 This certificate has no flags
FUNC_EXIT:  _iot_tls_verify_cert L#88 Return Code : 0 (../platform/mbedtls/network_mbedtls_wrapper.cpp)
FUNC_ENTRY: mbedtls_aws_send L#192 (../platform/mbedtls/network.cpp)
FUNC_EXIT:  mbedtls_aws_send L#204 Return Code : 876 (../platform/mbedtls/network.cpp)
ERROR: iot_tls_connect L#280  failed
  ! mbedtls_ssl_handshake returned -0x10 (-16)

FUNC_EXIT:  iot_tls_connect L#286 Return Code : -4 (../platform/mbedtls/network_mbedtls_wrapper.cpp)
FUNC_EXIT:  _aws_iot_mqtt_internal_connect L#404 Return Code : -4 (../src/aws_iot_mqtt_client_connect.c)
FUNC_ENTRY: iot_tls_disconnect L#408 (../platform/mbedtls/network_mbedtls_wrapper.cpp)
FUNC_EXIT:  iot_tls_disconnect L#416 Return Code : 0 (../platform/mbedtls/network_mbedtls_wrapper.cpp)
FUNC_ENTRY: iot_tls_destroy L#422 (../platform/mbedtls/network_mbedtls_wrapper.cpp)
FUNC_ENTRY: mbedtls_aws_free L#212 (../platform/mbedtls/network.cpp)
FUNC_EXIT:  mbedtls_aws_free L#220 (../platform/mbedtls/network.cpp)
FUNC_EXIT:  iot_tls_destroy L#434 Return Code : 0 (../platform/mbedtls/network_mbedtls_wrapper.cpp)
FUNC_ENTRY: aws_iot_mqtt_set_client_state L#110 (../src/aws_iot_mqtt_client.c)
FUNC_ENTRY: aws_iot_mqtt_get_client_state L#63 (../src/aws_iot_mqtt_client.c)
FUNC_EXIT:  aws_iot_mqtt_get_client_state L#68 Return Code : 2 (../src/aws_iot_mqtt_client.c)
FUNC_EXIT:  aws_iot_mqtt_set_client_state L#135 Return Code : 0 (../src/aws_iot_mqtt_client.c)
FUNC_EXIT:  aws_iot_mqtt_connect L#489 Return Code : -4 (../src/aws_iot_mqtt_client_connect.c)
FUNC_EXIT:  aws_iot_shadow_connect L#132 Return Code : -4 (../src/aws_iot_shadow.c)
ERROR: aws_shadow_sample_task L#167 Shadow Connection Error
 - - - - - - - ALL DONE - - - - - - -

**Comment by shokuie** [comment]
Any idea from experienced folks? I appreciate any hints.

**Comment by NikitaPhoghat** [comment]
It looks like certificate validation error. Did you have the system time set correctly on your ucleo-l476rg board.

**Comment by shokuie** [comment]
Thanks, @NikitaPhoghat for your reply, I tried setting the time in the main function but the result didn't change and here is the new output (without debug output):

AWS ../examples/shadow_sample/shadow_sample.cpp Example.

AWS IoT SDK Version 3.0.0-

Time as a basic string = Wed Jul 31 14:50:26 2019
Shadow Init
Shadow Connect
[EasyConnect] Using BG96
[EasyConnect] Connected to Network successfully
[EasyConnect] MAC address F2:48:61:00:20:00:20
[EasyConnect] IP address 25.45.203.69
ERROR: iot_tls_connect L#280  failed
  ! mbedtls_ssl_handshake returned -0x10 (-16)

ERROR: aws_shadow_sample_task L#173 Shadow Connection Error
 - - - - - - - ALL DONE - - - - - - -

I used OpenSSL and looks like it worked with the same certificates. So i have no clue, my hunch is an error caused by a difference between mbedtls and OpenSSL and I don't know what it is.

Would appreciate if you can give some more ideas.

**Comment by abhidixi11** [comment]
Hello shokuie,
I notice that the TLS handshake is failing in mbedtls library.  Are you connecting to AWS IOT core ? Did you generate new certificates? 
Can you get wire trace to verify if SSL handshake is going through correctly?

From the trace, it looks like you are using Avnet/aws-iot-device-sdk-mbed-c, which is fork of aws/aws-iot-device-sdk-mbed-c  and may contain older code. 
Thanks.

**Comment by shokuie** [comment]
Thanks @abhidixi11,

Yes, I'm connecting to AWS IOT Core and generated new ATS certificates. I'm using Avnet fork but their sample works perfectly with the predefined account using hardcoded certificates the project came with (the only difference between my code and the project is VeriSign CA and certificates), as soon as changing AWS account and replacing certificates with what I generated the error comes up.

As I mentioned OpenSSL and MQTT test client work (connect) perfectly using newly generated certificates from my Windows 10 Test machine. So most probably something is happening in mbedtls. 

For wire trace, as dev kit uses GSM connectivity it's not really possible to capture packets.

Regards

**Comment by abhidixi11** [comment]
Hello shokuie 

> As I mentioned OpenSSL and MQTT test client work (connect) perfectly using newly generated certificates from my Windows 10 Test machine. So most probably something is happening in mbedtls.

> FUNC_EXIT: mbedtls_aws_send L#204 Return Code : 876 (../platform/mbedtls/network.cpp)
> ERROR: iot_tls_connect L#280 failed

 The failure happens in network.cpp that you are using from Avnet/aws-iot-device-sdk-mbed-c . I have not really looked into details, but wondering if you can compare that with what we have in our latest version.

If you would like to continue using the fork, I would recommend enabling /adding more debug statements at TLS level to see what may be going wrong since you cannot get wire trace due to GSM.
Thanks.

**Comment by shokuie** [comment]
Thanks @abhidixi11 

Will take a deeper look and update you.

**Comment by abhidixi11** [comment]
OK Thanks,  I'm closing the issue, but if you find something that requires our attention, please let us know and feel free to reopen.

**Comment by shokuie** [comment]
Thank you @NikitaPhoghat for your hint about time issues. I remembered when porting the project to AC6 i got following linker error
`/../platform/common/timer.c:83: undefined reference to `timersub'`

then with some tricks and ifdef sorted it out but it looks like i messed timer functions. With in detail checks, I found this timer routine is used in gettimeofday functions that i believe is used when validating certificates. So i got back to the linker error and invested some time to understand I need to add linker flag -D_DEFAULT_SOURCE to include _BSD_SOURCE that has references for timersub and timeradd routines. And then everything worked like a charm. 

Thank you guys.
